libfwupdplugin/fu-device.c

```diff
@@ -657,7 +657,10 @@ fu_device_set_parent (FuDevice *self, FuDevice *parent)
 	if (parent != NULL &&
 	    fu_device_get_order (parent) == fu_device_get_order (self)) {
 		g_debug ("auto-setting %s order", fu_device_get_id (parent));
-		fu_device_set_order (parent, fu_device_get_order (self) + 1);
+		if (fu_device_has_flag (self, FWUPD_DEVICE_FLAG_INSTALL_PARENT_FIRST))
+			fu_device_set_order (self, fu_device_get_order (parent) + 1);
+		else
+			fu_device_set_order (parent, fu_device_get_order (self) + 1);
```

Comment:
The reason I didn't do this originally is that it's not going to work in the following scenario:

 * A is parent to B
 * B is parent to C

So if we have:

 * A added as parent to B (`A=>1, B=0`)
 * B added as parent to C (`B=>1, C=0`)

So, `A=1, B=1, C=0` which isn't what we want (we want `A=2, B=1, C=0`). I think you're on the right track, but I think we need to do a mini-depsolve for the entire composite device.

I can have a go at doing this on Monday if you want.

Reply:
https://github.com/fwupd/fwupd/pull/2387

Reply:
#2387 (plus the 3 first patches here) fixes the issue with install/activation order:

```
localhost ~ # fwupdtool update     
Loading?                 [***************************************]
Thunderbolt controller in Dell dock must remain connected for the duration of the update to avoid damage. Continue with update? [Y|n]: 
Decompressing?           [***************************************]
Installing on Package level of Dell dock?                        ]
Restarting device?       [***************************************]
Installing on VMM5331 in Dell dock?                              ]
Restarting device?       [***************************************] Less than one minute remaining?
Installing on WD19TB?    [    \                                  ]
Writing?                 [***************************************]
Installing on Thunderbolt controller in Dell dock?               ]
Writing?                 [***************************************]
Updates will complete after the dock is unplugged.  The dock will be unavailable for approximately 1 minute.
? Package level of Dell dock has the latest available firmware version
? RTS5413 in Dell dock has the latest available firmware version
? RTS5487 in Dell dock has the latest available firmware version
? VMM5331 in Dell dock has the latest available firmware version
? WD19TB has the latest available firmware version

localhost ~ # fwupdtool activate   
Loading?                 [***************************************]
Activating firmware update WD19TB?
Activating firmware update Thunderbolt controller in Dell dock?
```

Reply:
Thanks, that's great to hear.  I've rebased this branch on top of that and dropped last commit.